react/src/pages/utils/components/ChipSelect.tsx

```diff
@@ -0,0 +1,107 @@
+import React, { useEffect, useState } from "react";
+import { Chip, IconButton, makeStyles, Menu, MenuItem, Paper } from "@material-ui/core";
+import { AddCircle } from "@material-ui/icons";
+
+const useStyles = makeStyles(theme => ({
+    root: {
+        display: "flex",
+        justifyContent: "center",
+    },
+    grow: {
+        flexGrow: 1,
+    },
+    chipList: {
+        display: "flex",
+        justifyContent: "center",
+        flexWrap: "wrap",
+        listStyle: "none",
+        padding: theme.spacing(0.5),
+        margin: 0,
+    },
+    chip: {
+        margin: theme.spacing(0.5),
+    },
+}));
+
+interface SelectableChip {
+    label: string;
+    key: string;
+    selected?: boolean;
+}
+
+export default function ChipSelect(props: {
+    labels: string[];
+    selected?: string[];
+    onSelectionChange?: (selectedLabels: string[]) => void;
+}) {
+    const classes = useStyles();
+    const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
+    const [chips, setChips] = useState<SelectableChip[]>(
+        props.labels.sort().map(label => ({ label: label, key: label, selected: false }))
+    );
+    const [disableAdd, setDisableAdd] = useState(false);
+
+    useEffect(() => {
+        if (Array.isArray(props.selected)) {
+            setChips(chips =>
+                chips.map(chip => ({
+                    ...chip,
+                    selected: !!props.selected!.find(label => label === chip.label),
+                }))
+            );
+            setDisableAdd(props.selected.length === 0);
+        }
+    }, [props.selected]);
```

Comment:
This is all determined from the props, so is there a reason to run this in `useEffect` instead of just initializing the states from the props at the top?

Reply:
`useEffect` handles the case where the user clicks Cancel and reverts their changes

Reply:
Ah, I see